cmd/automerge-debug/main.go

```diff
@@ -0,0 +1,942 @@
+package main
+
+import (
+	"bytes"
+	"compress/flate"
+	"crypto/sha256"
+	"encoding/binary"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"strings"
+	"time"
+	"unicode/utf8"
+)
+
+// from: https://github.com/aviate-labs/leb128/blob/v0.1.0/leb.go#L13
+func leb128(n uint64) []byte {
+	leb := make([]byte, 0)
+	if n == 0 {
+		return []byte{0}
+	}
+	for n != 0x00 {
+		b := byte(n & 0x7F)
+		n >>= 7
+		if n != 0x00 {
+			b |= 0x80
+		}
+		leb = append(leb, b)
+	}
+	return leb
+}
+
+func readULEB(b []byte) (uint64, []byte, error) {
+	var n uint64
+	var i int
+	for {
+		if len(b) == 0 {
+			return 0, nil, fmt.Errorf("failed to find end of LEB")
+		}
+
+		c := b[0]
+		b = b[1:]
+		n |= uint64(c&0x7f) << (i * 7)
+		i++
+		if c&0x80 == 0 {
+			if c == 0 && i > 1 {
+				return n, b, fmt.Errorf("overly long LEB")
+			}
+			if i > 10 {
+				return n, b, fmt.Errorf("LEB > 64-bit")
+			}
+			if i == 10 && c&0x7e > 0 {
+				return n, b, fmt.Errorf("LEB > 64-bit")
+			}
+			return n, b, nil
+		}
+	}
+}
+
+func readSLEB(b []byte) (int64, []byte, error) {
+	var n uint64
+	var i int
+	var prev byte
+	for {
+		if len(b) == 0 {
+			return 0, nil, fmt.Errorf("failed to find end of LEB")
+		}
+
+		c := b[0]
+		b = b[1:]
+		n |= uint64(c&0x7f) << (i * 7)
+		i++
+		if c&0x80 == 0 {
+			if c&0x40 > 0 {
+				n |= (math.MaxUint64 << (i * 7))
+			}
+			if i > 1 && (prev&0x40 == 0 && c == 0 || prev&0x40 > 0 && c == 0x7f) {
+				return int64(n), b, fmt.Errorf("overly long LEB")
+			}
+			if i > 10 {
+				return int64(n), b, fmt.Errorf("LEB > 64-bit")
+			}
+			if i == 10 && c&0x7f != 0 && c&0x7f != 0x7f {
+				return int64(n), b, fmt.Errorf("LEB > 64-bit")
+			}
+			return int64(n), b, nil
+		}
+		prev = c
+	}
+}
+
+func addHeader(b []byte, kind byte) []byte {
```

Comment:
What about somiething like:

```go
var header = [4]byte{133, 111, 74, 131}

func addHeader(b []byte, kind byte) []byte {
	payload := append(leb128(uint64(len(b))), b...)
	hash := sha256.Sum256(payload)

	output := append([]byte{}, header[:]...)
	output = append(output, hash[:]...)
	output = append(output, kind)
	return output
}
```

- header is a well defined var (probably with a doc-comment)
- sum calculation is a well defined step
- and then just concat data (we can prealloc `output` slice 'cause we know it's final size)

WDYT ?

Reply:
Thanks! I've clarified the implementation a bit.